feature: AI-Enabled Food Search

taylorpatterson-T1D opened the pull request:
Initial commit

taylorpatterson-T1D committed:
Initial Commit

taylorpatterson-T1D commented:
Could I please get a review from the maintainers? @ps2, @marionbarker, @loudnate
 
This is my submission for a new AI-enabled food search capability for Loop. Loop has helped me become a better diabetic and I would love to give back to the community by contributing this new feature for Loop, which I believe will benefit all Loop users.

**PROBLEM WE'RE SOLVING:** Diabetics lack consistent nutrition knowledge when estimating carb intake. And we need that knowledge right in the moment, when we’re making dosing decisions - this knowledge is the single most important tool we have to improve our Time in Range and A1C. 

**NEW FEATURE'S IMPACT:** The Food Search system is a comprehensive food analysis and nutrition tracking solution integrated into Loop for improved diabetes management. It provides multiple search options including text, barcode scanning, voice, and AI-powered image analysis that produces comprehensive nutrition data and dosing suggestions via diabetic notes. This let's us enter a meal bolus with MUCH greater accuracy and improve Time in Range and A1C. The AI-enabled analysis engine can even read menu items from an image captured, and provide guidance in multiple languages!!!

Video Demo: https://youtu.be/L0LD8AxNX0Q

Review the docs in “//LoopWorkspace/Loop/Documentation/FoodSearch Docs” for details. API setup guides are provided so the user can enter their your own AI API keys for AI image analysis.

**New Search for Food section on the Add Carb Entry view:**
<img width="171" alt="1" src="https://github.com/user-attachments/assets/338fac62-12e3-49e7-9cb2-5173801c6f88" />

**Text search:**
<img width="174" alt="2" src="https://github.com/user-attachments/assets/21915cd3-1a26-4934-96e4-456dfc2322f8" />

**Barcode scan search for high accuracy:**
<img width="168" alt="3" src="https://github.com/user-attachments/assets/5c4fc5f2-33e2-4371-b8b8-935c1ebbd706" />

**AI-enabled food image analysis engine in action:**
<img width="172" alt="4" src="https://github.com/user-attachments/assets/a36fc95d-55c1-4330-a345-bde4d0c7c713" />

**AI-Analysis results with comprehensive analysis details including an item by item breakdown of the meal:**
<img width="170" alt="5" src="https://github.com/user-attachments/assets/8eb861c2-5f67-4e10-bf81-6b9b381ecbeb" />

**Even more detail about how Portions and Servings were calculated and diabetic specific notes to advise during dosing:**
<img width="167" alt="6" src="https://github.com/user-attachments/assets/dc660d3c-f5de-42a6-90d0-8e1835313ba3" />

**New Setup screen. Pick the AI service you prefer, enter your own API keys:**
<img width="169" alt="7" src="https://github.com/user-attachments/assets/03f4add0-ce21-417c-b563-8ef5a9cf95dc" />

taylorpatterson-T1D commented:
Video Demo is here: https://youtu.be/L0LD8AxNX0Q

taylorpatterson-T1D committed:
User Experience Improvements:

     ☐ Added network quality detection
     ☐ Implement safer timeout handling
     ☐ Add fast mode fallback logic

1. On Good WiFi: Fast parallel processing (unchanged performance)
2. On Restaurant WiFi: Sequential processing prevents freezing
3. On Cellular: Conservative mode with longer timeouts
4. On Poor Networks: Automatic fallback to safer processing

taylorpatterson-T1D committed:
AI Prompt Improvements

Improved AI prompt to include glycemic index analysis and strengthen diabetic notes section. Removed excessive DEBUG logging for better overall performance.

marionbarker commented on Loop.xcodeproj/project.pbxproj:
```diff
@@ -5146,6 +5288,7 @@
 				"OTHER_SWIFT_FLAGS[arch=*]" = "-DDEBUG";
 				"OTHER_SWIFT_FLAGS[sdk=iphonesimulator*]" = "-D IOS_SIMULATOR -D DEBUG";
 				PRODUCT_BUNDLE_IDENTIFIER = "$(MAIN_APP_BUNDLE_IDENTIFIER)";
+				"PRODUCT_BUNDLE_IDENTIFIER[sdk=iphoneos*]" = com.4S2EW2Q6ZW.loopkit.Loop;
```
Please remove this line. It breaks the ability for people to automatically sign with their developer ID (it inserts yours).  The line is not needed.

taylorpatterson-T1D replied:
done

marionbarker commented on Loop.xcodeproj/project.pbxproj:
```diff
@@ -5166,13 +5309,15 @@
 				);
 				OTHER_LDFLAGS = "";
 				PRODUCT_BUNDLE_IDENTIFIER = "$(MAIN_APP_BUNDLE_IDENTIFIER)";
+				"PRODUCT_BUNDLE_IDENTIFIER[sdk=iphoneos*]" = com.4S2EW2Q6ZW.loopkit.Loop;
```
Please remove this line. It breaks the ability for people to automatically sign with their developer ID (it inserts yours).  The line is not needed.

taylorpatterson-T1D replied:
done

marionbarker commented on Loop.xcodeproj/project.pbxproj:
```diff
@@ -3,7 +3,7 @@
 	classes = {
 	};
-	objectVersion = 54;
+	objectVersion = 70;
```
Does the objectVersion need to be raised? (This means Xcode 16.0 or later, whereas objectVersion of 54 is Xcode 12 or later.)

I don't know the answer to the question, but I did restore it to 54 in my local clone and it built and seemed to run the same.

taylorpatterson-T1D replied:
reverted to 54

taylorpatterson-T1D replied:
I did some research and testing on this. Because I'm using newer xCode it automatically updates this. Older xCode users will be unable to use my repo unless they manually make the change you did first, which appears safe to do since my updates do not leverage an 70-specific capabilities. My options are:
1. Always develop in the oldest xcode version supported
2. Clarify instructions stating one must manually update prior to build
3. Include manual adjustment in the build script we might create when attempting to add this to the Customizations build script
4. Provide instructions to update to latest version of xCode before building. (which LoopDocs already does).

When testing, it did NOT add back my Dev Team details. That is happening when I build on my machine for some reason. Still tracking down why and will fix that if I can.

marionbarker replied:
When you make the other changes I requested to the branch, this will no longer be automatically modified and your Apple Developer ID will not be inserted into the pbxproj file - as happened again with this commit.

I am testing the rest of your most recent commit, 2adc9a3e, and will comment on it separately. But the changes to `Loop.xcodeproj/project.pbxproj` from this one commit need to be reverted.

marionbarker reviewed (changes requested):
This is the first of several comments. It includes 2 specific lines that need to be removed from `Loop.xcodeproj/project.pbxproj`.  The 2 lines contain Taylor's Developer ID. They break the automatic signing configuration for Loop and are not required - they can be deleted.



In addition to those 2 lines in `Loop.xcodeproj/project.pbxproj`, I do not know why the LibreTransmitter information was touched. I think that is a mistake. (It might not harm anything, but we prefer not to make changes unless they are necessary.)

marionbarker commented:
More general comments:

This was pretty slick. The bar code reader was the one I played with the most. The general food search seemed to spend a long time loading, but maybe my network was having an issue.

I will do more testing and report back.

Regarding this addition, I'm not sure of the best way to incorporate it. I'm sure it is user specific whether they want this level of detail or if they go with a few Favorite Food listings and use a SWAG for the carbs to enter (which is my approach). We might want to move the position of the food search/service lower on the screen into its own FOOD SEARCH area (like FAVORITE FOODS).

You might want to upload the images in your comment as png files - they are distorted and hard to read on my phone and distorted and impossible to read on my computer browser. 

The following graphic was made as a composite from my test phone (an SE 2nd gen) - the display is taller than will fit on the phone with the keyboard deployed.


![pr2325-food-search](https://github.com/user-attachments/assets/a073e93f-cf7d-41e2-b8dd-3b1e92d4d8d8)

taylorpatterson-T1D commented:
@marionbarker  -  Thank you for your kind help and suggestion. I have repaired the Taylor-specific issues and reverted the version uptick.

If we make this functionality a new Customization option from the main Loop build script then the user is saying that they want to prioritize the value delivered by making food search results their new priority (versus anything they may have used in the past).  So keeping the functionality at the top of the Add Carb Entry view seems appropriate.

If they don't like the functionality then it can be removed returning Add Carb Entry to its default. 

If they make the choice to add Food Search, then my POV is that it should remain at the top of the Add Carb Entry section and not be moved below (near Favorite Foods) where it may go overlooked and therefore unused.

Your thoughts?

taylorpatterson-T1D committed:
Removed developer-specific team refs in Loop.xcodeproj/project.pbxproj

Also reverted object version to 54

taylorpatterson-T1D committed:
Merge branch 'dev' of https://github.com/taylorpatterson-T1D/Loop into dev

taylorpatterson-T1D committed:
Repaired object version (again)

marionbarker reviewed (changes requested):
I did another review and there are still some changes that need to be removed.

Please note: this only builds with **Mac-Xcode**. I am unable to build with GitHub Actions (at this time). I tested both the version in PR 2325, commit feb6f262, and the version with the items removed as described below.

For GitHub errors, please see: https://github.com/docs-test/LoopWorkspace/actions/runs/16239378975

## Modification Summary

I removed these items and tested that Loop with food search still builds correctly with Mac (macOS 15.5) and Xcode (16.3):

1. I removed all references to LibreTransmitter from `Loop.xcodeproj/project.pbxproj`
2. I removed `Loop.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved`
3. I removed the entire folder `Loop/Loop.xcodeproj` (the Loop.xcodeproj folder should only be at the top level, this was duplicated at a lower level)

I pushed my version to the loopandlearn github account so I could link some of the changes in the details below and to support testing the build with GitHub actions.

## Modification Details

Before making these changes
* Be sure to clean build folder and close the workspace in Xcode
* Only open Xcode workspace again after changes are made

### Item 1:

see this commit: https://github.com/loopandlearn/Loop/commit/f2874d25348d11a163e3b2ccfe8ebd5001d97692 

### Item 2:
command line to take this action:
```
% git rm Loop.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved
rm 'Loop.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved'
```
see this commit: https://github.com/loopandlearn/Loop/commit/aa67ab8ba47ffe24b2cb64d30d6cc13a69ac1535

### Item 3:
command line to take this action:
```
% git rm -r Loop/Loop.xcodeproj 
rm 'Loop/Loop.xcodeproj/project.pbxproj'
rm 'Loop/Loop.xcodeproj/project.xcworkspace/contents.xcworkspacedata'
rm 'Loop/Loop.xcodeproj/project.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist'
rm 'Loop/Loop.xcodeproj/project.xcworkspace/xcshareddata/WorkspaceSettings.xcsettings'
rm 'Loop/Loop.xcodeproj/xcshareddata/xcschemes/DoseMathTests.xcscheme'
rm 'Loop/Loop.xcodeproj/xcshareddata/xcschemes/Loop Intent Extension.xcscheme'
rm 'Loop/Loop.xcodeproj/xcshareddata/xcschemes/Loop Status Extension.xcscheme'
rm 'Loop/Loop.xcodeproj/xcshareddata/xcschemes/Loop.xcscheme'
rm 'Loop/Loop.xcodeproj/xcshareddata/xcschemes/LoopTests.xcscheme'
rm 'Loop/Loop.xcodeproj/xcshareddata/xcschemes/SmallStatusWidgetExtension.xcscheme'
rm 'Loop/Loop.xcodeproj/xcshareddata/xcschemes/WatchApp.xcscheme'
```

see this commit: https://github.com/loopandlearn/Loop/commit/b40a22098b4600652636f2f7564d13a54da6078c

marionbarker commented:
Regarding not being able to build with GitHub actions. The modification in this link fixed the error.
* https://github.com/loopandlearn/Loop/commit/b40cb3c91c4b29173c3e253ce07b80b6ae91102e

Required change:
* add missing arguments as indicated in the error message

Other change:
* be consistent using imageUrl instead of imageURL in all arguments and labels
   *  the imageURL version was found in some places with imageUrl in most places

Error reported using Xcode: Product->Archive was:
```
/Users/marion/Downloads/ManualClones/Loop_dev_updates/LoopWorkspace/
 Loop/Loop/Services/FoodSearchRouter.swift:
 240:58: error: missing arguments for parameters
 'servingQuantity', 'imageFrontUrl', 'code' in call

            let placeholderProduct = OpenFoodFactsProduct(
```

marionbarker commented:
In order for this to be considered for adding to Loop (instead of as a customization - which is a different process), I think the Loop, Settings, AI Food Analysis row would need an addition for showing or not showing the AI tool in the carb screen. Following other additions to the UI, the default would be for this to be disabled.

<img width="1673" height="1729" alt="loop-pr2325_suggestion" src="https://github.com/user-attachments/assets/80d7ac75-a077-460e-a4d1-9a9de9719669" />

taylorpatterson-T1D commented:
@marionbarker 
Easy peasy. I will add that and send another pull request tonight or tomorrow. 

On the subject of the project files stuff, I was unable to build by following your directions of removing those sub contents. It wouldn't see the Loop project as valid without them, unless I'm doing something wrong.

taylorpatterson-T1D commented:
Feature Flag Implementation Done @marionbarker 
Files Modified:
  1. UserDefaults+Loop.swift:
    - Added foodSearchEnabled key to enum
    - Added computed property with default value false (disabled by default)
    - Settings are persistent - when disabled, all AI/provider settings remain saved
  2. AISettingsView.swift:
    - Added feature toggle at the top of the settings screen
    - Wrapped all configuration sections in if foodSearchEnabled { }
    - Medical disclaimer remains always visible
    - Clear messaging about feature control
  3. CarbEntryView.swift:
    - Wrapped entire Food Search UI section with feature flag check
    - Properly handles CardSectionDivider logic for both enabled/disabled states
    - Only shows when isNewEntry && UserDefaults.standard.foodSearchEnabled
  4. SettingsView.swift:
    - Updated descriptive text to clarify this enables/configures Food Search

  User Experience:
  Default State: Food Search is disabled by default for new users Settings Persistence: All AI API keys and
  configurations remain saved when disabled. Easy Toggle: Users can enable/disable via toggle in Food Search settings. Clean UI: When disabled, the red-boxed UI section completely disappears. No Data Loss: Re-enabling restores all previous
  settings

  Behavior:
  - Disabled: Food Search section hidden from carb entry screen
  - Enabled: Full Food Search functionality appears (search bar, AI camera, barcode scanner, voice search)
  - Settings: AI Settings always accessible so users can toggle the feature on/off

taylorpatterson-T1D committed:
Make Food Search a feature flag so it can be included in main branch.

Move the Food Search behind a feature flag found in Settings
Improved AI prompt to prevent overdose recommendations
Streamlined Diabetic Notes, Portion Details and other UX improvements

taylorpatterson-T1D commented:
Food Search Feature Flag Implementation Done!
Files Modified:
UserDefaults+Loop.swift:
- Added foodSearchEnabled key to enum
- Added computed property with default value false (disabled by default)
- Settings are persistent - when disabled, all AI/provider settings remain saved
AISettingsView.swift:
- Added feature toggle at the top of the settings screen
- Wrapped all configuration sections in if foodSearchEnabled { }
- Medical disclaimer remains always visible
- Clear messaging about feature control
CarbEntryView.swift:
- Wrapped entire Food Search UI section with feature flag check
- Properly handles CardSectionDivider logic for both enabled/disabled states
- Only shows when isNewEntry && UserDefaults.standard.foodSearchEnabled
SettingsView.swift:
- Updated descriptive text to clarify this enables/configures Food Search
User Experience:
Default State: Food Search is disabled by default for new users Settings Persistence: After the user enables Food Search and adds any provider choices or API keys for AI Analysis, all AI API keys and configurations remain saved when later disabled again. Easy Toggle: Users can enable/disable via toggle in Food Search settings. Clean UI: When disabled, the Search for Food UI section completely disappears. No Data Loss: Re-enabling restores all previous settings

I also improved the AI prompt and context to add Glycemic Index inputs, and refined how the food is presented in the UI to make it easier to read.

@marionbarker , @ps2 
Since it's now hidden behind a feature flag configurable by the user, can we add this functionality to the main branch of Loop???

<img width="354" height="769" alt="Screenshot 2025-07-13 at 5 47 46 PM" src="https://github.com/user-attachments/assets/de4527ca-4eda-497b-9463-1619c7ba4a7e" />

<img width="354" height="545" alt="Screenshot 2025-07-13 at 5 47 56 PM" src="https://github.com/user-attachments/assets/e18e091a-5355-4ffd-8d83-9eb65a1de65f" />

<img width="356" height="772" alt="Screenshot 2025-07-13 at 5 48 05 PM" src="https://github.com/user-attachments/assets/d2e87b92-deaa-48cb-8303-a5790ac44bf5" />

marionbarker commented:
General comments about this commit: [2adc9a3](https://github.com/LoopKit/Loop/pull/2325/commits/2adc9a3e23251356b2ff5bde442a4c9ca816c796):

I have not finished my detailed review and testing yet, but wanted to get this general comment out.

I already mentioned the changes to `Loop.xcodeproj/project.pbxproj` need to be reverted.

You did not implement some of my earlier suggestions because you "were unable to build" - we can talk about those on zulipchat via DM if you prefer.

The other general comment is that it is better to do atomic commits that change one thing at a time. Please consider this for future commits - I'm not asking you to change this one.

I see 3 "changes" in this one commit:
1. You modified `imageUrl` to be `imageURL`
   * This is opposite from the choice I made but I'm not sure if it matters one way or the other, I will confirm that it is now consistent in all places in a later review
2. You added features to speed up the process
3. You added the requested feature to add an enable/disable switch to display the AI tool on the Carb screen

marionbarker commented:
## Summary

The modifications I previously requested are still required before this can be considered.
* With those modifications made, this successfully builds both with Mac-Xcode and with Browser Build.
* Thank you for adding the enable/disable flag.

Once you make the required modifications, I will test again and then we can see if @ps2 will approve it.

Be sure the modifications are made with Xcode closed. Then when you open Xcode again, do a clean build folder, close the workspace and then open it again. If you do not take those steps, you will get build failures.

## Modifications

I started with a fresh branch (`modify-taylor-dev-pr2325_vs3`) using your latest commit, 2adc9a3e, and repeated most of the changes I had previously made (in earlier testing) using cherry-pick. Some had to be handled differently as described below. (The first 3 commits were cherry-pick, the last 3 commits were manual.)

You can see the commits I applied to your version:
* https://github.com/loopandlearn/Loop/commits/modify-taylor-dev-pr2325_vs3/

The successful Browser Build is:
* https://github.com/docs-test/LoopWorkspace/actions/runs/16272023435

### Modification Details

One of the things you did not add was the extra parameters required to enable Browser Build (or Product -> Archive) in `Loop/Services/FoodSearchRouter.swift`.

The missing arguments I did manually (instead of  cherry-pick) because of the case you chose for imageURL.

It turns out that imageFrontUrl must be spelled with Url in title case, so it took me 2 tries to get that modification correct. 

**Details about case**

Initially, you had some instances of `imageUrl` and others that used `imageURL`:
* You chose to change all instances to `imageURL`
* I had chosen the other way because `imageUrl` was more frequent in the code at the time
* So now all `imageURL` instances use upper case (capital letters for URL) and all `imageFrontUrl` use title case
* This does not prevent the code from building and operating as expected but it can lead to someone typing the wrong thing

taylorpatterson-T1D commented:
> I did another review and there are still some changes that need to be removed.
> 
> Please note: this only builds with **Mac-Xcode**. I am unable to build with GitHub Actions (at this time). I tested both the version in PR 2325, commit [feb6f26](https://github.com/LoopKit/Loop/commit/feb6f26219517382ef9b474a62f197ffc8121a5e), and the version with the items removed as described below.
> 
> For GitHub errors, please see: https://github.com/docs-test/LoopWorkspace/actions/runs/16239378975
> 
> ## Modification Summary
> I removed these items and tested that Loop with food search still builds correctly with Mac (macOS 15.5) and Xcode (16.3):
> 
> 1. I removed all references to LibreTransmitter from `Loop.xcodeproj/project.pbxproj`
> 2. I removed `Loop.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved`
> 3. I removed the entire folder `Loop/Loop.xcodeproj` (the Loop.xcodeproj folder should only be at the top level, this was duplicated at a lower level)
> 
> I pushed my version to the loopandlearn github account so I could link some of the changes in the details below and to support testing the build with GitHub actions.
> 
> ## Modification Details
> Before making these changes
> 
> * Be sure to clean build folder and close the workspace in Xcode
> * Only open Xcode workspace again after changes are made
> 
> ### Item 1:
> see this commit: [loopandlearn@f2874d2](https://github.com/loopandlearn/Loop/commit/f2874d25348d11a163e3b2ccfe8ebd5001d97692)
> 
> ### Item 2:
> command line to take this action:
> 
> ```
> % git rm Loop.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved
> rm 'Loop.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved'
> ```
> 
> see this commit: [loopandlearn@aa67ab8](https://github.com/loopandlearn/Loop/commit/aa67ab8ba47ffe24b2cb64d30d6cc13a69ac1535)
> 
> ### Item 3:
> command line to take this action:
> 
> ```
> % git rm -r Loop/Loop.xcodeproj 
> rm 'Loop/Loop.xcodeproj/project.pbxproj'
> rm 'Loop/Loop.xcodeproj/project.xcworkspace/contents.xcworkspacedata'
> rm 'Loop/Loop.xcodeproj/project.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist'
> rm 'Loop/Loop.xcodeproj/project.xcworkspace/xcshareddata/WorkspaceSettings.xcsettings'
> rm 'Loop/Loop.xcodeproj/xcshareddata/xcschemes/DoseMathTests.xcscheme'
> rm 'Loop/Loop.xcodeproj/xcshareddata/xcschemes/Loop Intent Extension.xcscheme'
> rm 'Loop/Loop.xcodeproj/xcshareddata/xcschemes/Loop Status Extension.xcscheme'
> rm 'Loop/Loop.xcodeproj/xcshareddata/xcschemes/Loop.xcscheme'
> rm 'Loop/Loop.xcodeproj/xcshareddata/xcschemes/LoopTests.xcscheme'
> rm 'Loop/Loop.xcodeproj/xcshareddata/xcschemes/SmallStatusWidgetExtension.xcscheme'
> rm 'Loop/Loop.xcodeproj/xcshareddata/xcschemes/WatchApp.xcscheme'
> ```
> 
> see this commit: [loopandlearn@b40a220](https://github.com/loopandlearn/Loop/commit/b40a22098b4600652636f2f7564d13a54da6078c)

taylorpatterson-T1D committed:
Delete Loop/DefaultAssets.xcassets/AI-logo-master.png

taylorpatterson-T1D committed:
Update project.pbxproj

taylorpatterson-T1D committed:
Update project.pbxproj to remove Taylor's team ID

taylorpatterson-T1D committed:
Update project.pbxproj - cleaned unnecessary LibreTransmitter code

taylorpatterson-T1D committed:
Update Package.resolved

taylorpatterson-T1D committed:
Delete redundant Loop.xcodeproj directory

taylorpatterson-T1D committed:
Delete Loop/Loop.xcodeproj/project.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist

taylorpatterson-T1D committed:
Delete Loop/Loop.xcodeproj/project.xcworkspace/xcshareddata/WorkspaceSettings.xcsettings

taylorpatterson-T1D committed:
Delete Loop/Loop.xcodeproj/project.xcworkspace/contents.xcworkspacedata

taylorpatterson-T1D committed:
Delete Loop/Loop.xcodeproj/xcshareddata/xcschemes/DoseMathTests.xcscheme

taylorpatterson-T1D committed:
Delete Loop/Loop.xcodeproj/xcshareddata/xcschemes/Loop Intent Extension.xcscheme

taylorpatterson-T1D committed:
Delete Loop/Loop.xcodeproj/xcshareddata/xcschemes/Loop Status Extension.xcscheme

taylorpatterson-T1D committed:
Delete Loop/Loop.xcodeproj/xcshareddata/xcschemes/WatchApp.xcscheme

taylorpatterson-T1D committed:
Delete Loop/Loop.xcodeproj/xcshareddata/xcschemes/SmallStatusWidgetExtension.xcscheme

taylorpatterson-T1D committed:
Delete Loop/Loop.xcodeproj/xcshareddata/xcschemes/LoopTests.xcscheme

taylorpatterson-T1D committed:
Delete Loop/Loop.xcodeproj/xcshareddata/xcschemes/Loop.xcscheme

taylorpatterson-T1D commented:
> rm 'Loop/Loop.xcodeproj/project.pbxproj'

@marionbarker 
I did all the reset of the deletions but before I do this one - I wanted to clarify, won't this result in Loop not having all the additional Food Search related files registered in the project? Would that force the builder to have to re-add them to their own project.pbxproj?

taylorpatterson-T1D commented:
> Regarding not being able to build with GitHub actions. The modification in this link fixed the error.
> 
> * [loopandlearn@b40cb3c](https://github.com/loopandlearn/Loop/commit/b40cb3c91c4b29173c3e253ce07b80b6ae91102e)
> 
> Required change:
> 
> * add missing arguments as indicated in the error message
> 
> Other change:
> 
> * be consistent using imageUrl instead of imageURL in all arguments and labels
>   
>   * the imageURL version was found in some places with imageUrl in most places
> 
> Error reported using Xcode: Product->Archive was:
> 
> ```
> /Users/marion/Downloads/ManualClones/Loop_dev_updates/LoopWorkspace/
>  Loop/Loop/Services/FoodSearchRouter.swift:
>  240:58: error: missing arguments for parameters
>  'servingQuantity', 'imageFrontUrl', 'code' in call
> 
>             let placeholderProduct = OpenFoodFactsProduct(
> ```

I chose imageURL as the standard for consistency because URL is an acronym and therefore more likely to be more future proof if others contribute to the feature. This was tested as working on my end so it should now be resolved with my last commit.

marionbarker commented:
@taylorpatterson-T1D 

Please do not be discouraged. But please read carefully.

### commit you need to revert 

You labeled this commit as "Delete redundant Loop.xcodeproj directory"
* https://github.com/LoopKit/Loop/pull/2325/commits/48e88e07dcaaffe1b3b99ef2c81422d41cbf5ccd

But in fact you deleted the one directory that was correct, i.e., you deleted the `Loop.xcodeproj` folder you were supposed to keep. So of course it won't build, in fact you get an error just by opening the workspace.

I tried fixing just that commit with this command:
```
git revert 48e88e07
```

And in my clone it now builds fine.

### atomic commits

When I said use atomic commits - I meant all changes in a single commit should solve a problem. It might be a lot of files. Doing this is not required but it is a good idea because then reviewers can look at the changes made per commit. It makes it easier to review.

### test build before push

Please do not push any commits until you have a working build. Then push them to GitHub at one time. 

When you pushed following every commit of deleting single files (all for one purpose) yesterday, that sent a separate email to every one who watches the Loop repository.

You deleted (one file at a time) items found in the `Loop/Loop.xcodeproj` directory that was truly redundant. But you did not delete the directory itself.

### missing arguments

You have not fixed the missing arguments that will prevent this from working for a browser build.

Test this yourself by taking these actions in Xcode:
* Product, Clean Build Folder
* Product, Archive

The build will fail. Click on the Report Navigator icon
Click on the error:

Showing Recent Messages
/Users/marion/Downloads/ManualClones/Loop_dev_updates/LoopWorkspace/Loop/Loop/Services/FoodSearchRouter.swift:240:58: Missing arguments for parameters 'servingQuantity', 'imageFrontUrl', 'code' in call

marionbarker commented:
Found another must-fix issue:

## Conflicts with Favorite Foods

**build that includes just this customization**
* With the food search added to the build: I can select a stored Favorite Food but the settings for everything, carb amount, Food Type, Absorption time are not updated on the carb entry screen
* This happened whether the food_search feature was enabled or disabled 

**build main 3.6.3**
* Returning to main without food search added: I can select from my stored favorite foods and all saved settings including carb amount, Food Type, Absorption time are properly entered

This conflict was evaluated on 3 different phones and found to happen on all of them:
* iPhone 8 running iOS 16.7.11
* SE 2nd gen running iOS 18.5
* iPhone 15 pro running iOS 18.5

taylorpatterson-T1D commented:
> @taylorpatterson-T1D
> 
> Please do not be discouraged. But please read carefully.
> 
> ### commit you need to revert
> You labeled this commit as "Delete redundant Loop.xcodeproj directory"
> 
> * [48e88e0](https://github.com/LoopKit/Loop/commit/48e88e07dcaaffe1b3b99ef2c81422d41cbf5ccd)
> 
> But in fact you deleted the one directory that was correct, i.e., you deleted the `Loop.xcodeproj` folder you were supposed to keep. So of course it won't build, in fact you get an error just by opening the workspace.
> 
> I tried fixing just that commit with this command:
> 
> ```
> git revert 48e88e07
> ```
> 
> And in my clone it now builds fine.
> 
> ### atomic commits
> When I said use atomic commits - I meant all changes in a single commit should solve a problem. It might be a lot of files. Doing this is not required but it is a good idea because then reviewers can look at the changes made per commit. It makes it easier to review.
> 
> ### test build before push
> Please do not push any commits until you have a working build. Then push them to GitHub at one time.
> 
> When you pushed following every commit of deleting single files (all for one purpose) yesterday, that sent a separate email to every one who watches the Loop repository.
> 
> You deleted (one file at a time) items found in the `Loop/Loop.xcodeproj` directory that was truly redundant. But you did not delete the directory itself.
> 
> ### missing arguments
> You have not fixed the missing arguments that will prevent this from working for a browser build.
> 
> Test this yourself by taking these actions in Xcode:
> 
> * Product, Clean Build Folder
> * Product, Archive
> 
> The build will fail. Click on the Report Navigator icon Click on the error:
> 
> Showing Recent Messages /Users/marion/Downloads/ManualClones/Loop_dev_updates/LoopWorkspace/Loop/Loop/Services/FoodSearchRouter.swift:240:58: Missing arguments for parameters 'servingQuantity', 'imageFrontUrl', 'code' in call

Fixed in Food Search 2.0 (new PR coming): The error should now be resolved since we're using the complete initializer signature that matches the definition in OpenFoodFactsModels.swift:145. The dataSource: .aiAnalysis is appropriate since these are AI-generated food entries from text searches. Running Product > Archive now works without error.

taylorpatterson-T1D commented:
> Fixed in Food Search 2.0 (new PR coming): The error should now be resolved since we're using the complete initializer signature that matches the definition in OpenFoodFactsModels.swift:145. The dataSource: .aiAnalysis is appropriate since these are AI-generated food entries from text searches. Running Product > Archive now works without error.

marionbarker commented:
Please see #2329 for newer work.
